### ISSUE TITLE
Making slurm logs belong to the fractal user

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-server#317]
Jumping on the work here: https://github.com/fractal-analytics-platform/fractal-server/issues/315 and extending on the discussion here: https://github.com/fractal-analytics-platform/fractal-server/issues/292

Is it possible to make the log files being owned by the fractal user?

This came up as part of the discussion around the deployment of Fractal at FMI today. Given that we're now more in control of those files with our own backend, having these mixed permissions can lead to some issues (i.e. the fractal user not being able to delete them).

Could we make those files be owned by the fractal user to avoid this?

We would still need to give the users access to write in them while the workflow is running (but could actually change the permissions when the workflow finishes to avoid users modifying or deleting log files). Enrico suggested the fractal user could create empty log files (touch filename.err) and give the users permission to write to them.
[the removing of write access later seems like an extra feature to me]

**Comment by jluethi** [comment]
Notes from today's discussion:
Doing this directly will be tricky (not tested extensively). To avoid this being a blocking issue and problems removing logs as fractal user, we can set the permissions on the log files broadly for the time being.
One workaround if we need to achieve ownership by fractal user will be copying over the logs to a new, fractal-owned folder after the processing has finished. Adds extra complexity though, so let's see if it's needed.

**Comment by jluethi** [comment]
And another idea:
Using `setfacl` to give the specific user that's running a job access to just that folder
i.e.:

> Fractal would need to set the correct ACL for each job folder for the correct user. At least this would be the best solution. If easier you can bodge it by moving the artifacts (and any other folder the user need to write into) to /tmp/<fractal-subfolder> and then copy them over once the job is done. But honestly the first option sounds much simpler as you need to add pretty much a single line of code to fractal, something like subprocess.run(['setfacl', blablabla])

**Comment by tcompa** [comment]
We had a look at the `setfacl` plan, with @mfranzon.

TL;DR
* It seems to "work" for our intended use (note: ownerships are not modified, but access is much more strict).
* Its vanilla version does not work on the nfs4 UZH filesystem, because it would require us to use `nfs4_setfacl` command.

---

### Preliminary: which filesystem?

* We work in the `/tmp` folder.
* `df /tmp` shows that it is part of `/dev/vda1`
* `mount | grep "/dev/vda1"` shows that this is an `ext4` filesystem.
* On the contrary, using `/data/homes/fractal` as a folder points to a `nfs4` filesystem. We discard that case by now.

### How to reproduce

As `fractal` user
```
$ mkdir /tmp/w5

$ chmod 700 /tmp/w5

$ ls -la /tmp/w5 
total 6172
drwx------ 2 fractal fractal    4096 Dec 15 10:03 .

$ setfacl -m u:test01:rwx,u:fractal:rwx /tmp/w5

$ getfacl /tmp/w5
getfacl: Removing leading '/' from absolute path names
# file: tmp/w5
# owner: fractal
# group: fractal
user::rwx
user:test01:rwx
user:fractal:rwx
group::---
mask::rwx
other::---
```

As `test01` user
```
echo "this is a log" > /tmp/w5/log.txt
```

As `fractal` user, this command works
```
rm /tmp/w5/log.txt -f
```


### What to do about nfs4 filesystems?

In principle `nfs4_setfacl` should correspond to `setfacl`, but we would need to test it.

### How large a change for fractal-server?

Code-wise, a few small changes:
1. We need to expose a variable that knows something about the filesystem (e.g. ext4 or nfs4)
2. We need to remove all 777 permissions
3. We need to add a setfacl command right after a workflow dir is created

On a higher level, we are adding a hard requirement to fractal, in terms of which filesystems are supported (and these filesystems need be configured with ACL). At the moment it's not obvious whether this is a severe restriction.

### Another important aspect to consider
This was all for the workflow folders. What about the tasks-environment ones? Let's look at it more carefully.

**Comment by tcompa** [comment]
Here is what we would by only using chmod

As fractal user
```
mkdir /tmp/w6
chmod 777 /tmp/w6
```

As test01 user
```
echo "This is a log" > /tmp/w6/log
```

As fractal user (**to be done at the end of the workflow execution**):
```
chmod 700 /tmp/w6
```

And now this command (as a fractal user) works:
```
rm /tmp/w6/log
```

The issue here is that during the execution of the workflow the folder would be broadly accessible, and the permissions would only be restricted at the end.

**Comment by jluethi** [comment]
Thanks a lot @tcompa & @mfranzon !

After some internal discussion, we need to rule out the chmod option, but got another suggestion to evaluate vs. the ACL approach. Here's are the relevant messages from Enrico:

> 777 is never an option unfortunately. It would introduce a race condition and it's the "perfect juicy target". Bad idea. ACLs are better, but there is in fact one very infamous case where they don't work: NFSv4, not with POSIX ACLs anyway. Any other file system (including NFSv3 provided the server is Linux OS) supports POSIX ACLs. At FMI we don't use NFSv4 or NFS ACLs as we use POSIX ACLs everywhere. I do not recommend the use of NFSv4 ACLs because standard tools do not support them (e.g. rsync). When you copy your data your NFS ACL will be lost, which is not ideal (including backup).
> 
> The best and most generic solution is the use of IPC / network messages or streaming via SLURM. No ACL or permission change required, just authentication between the two IPC ends (easily provided by a service such as munge). The resulting code becomes much easier to manage as you don't have to pay attention to permission state and timing any longer. I warmly recommend switching away from permission, the more I think about it, the more wrong it feels.

And on the discussion of making it more easy to be used broadly:
> Making Fractal easier to run on any setup it even more of a reason to stop looking at permissions. IPC  and network messages / streaming are a much better solution and should be more portable to other setups. Permissions and co. can be restricted or different depending on the setup and this kind of changes might not work somewhere else. And it's a security nightmare.

Could you look into comparing the ACL approach with an approach using IPC / network messages or streaming via SLURM?

**Comment by tcompa** [comment]
Reporting about the other option we had mentioned (UNIX groups):
```
As admin
sudo adduser fractal
sudo adduser test01
sudo usermod -aG fractal test01

As fractal
mkdir /tmp/workingdir

As test01
echo "i was here" > /tmp/workingdir/log

As fractal, you cannot modify the file:
touch /tmp/workingdir/log # this leads to a "Permission denied", 

As fractal, you can copy/move/remove the file. All the following commands work
cp /tmp/wd/log /tmp/workingdir/log2
mv /tmp/wd/log /tmp/workingdir/log3
rm /tmp/workingdir/log
```

Note that a third user who is not in the `fractal` group will not be able to `rm/mv` the log file (they can still `cp` it), but other users in the `fractal` group will be able to do it (as the `fractal` user can).

**Comment by tcompa** [comment]
These were the two options we discussed with @jluethi and @gusqgm

### Option A
We use ACLs, as in https://github.com/fractal-analytics-platform/fractal-server/issues/317#issuecomment-1352836505.
First steps: find and test the nfs4  `setfacl`-equivalent command; make sure that calling these commands (`setfacl` or equivalent) from python subprocessing works fine (it should); have an env variable within fractal-server to know which filesystem we are on.

### Option B
We proceed with UNIX groups, as in https://github.com/fractal-analytics-platform/fractal-server/issues/317#issuecomment-1357503304, but only during workflow execution. At the end of a workflow execution (be it a success or a failure), we "lock" the folder by (1) copying all user-owned files to fractal-owned files, (2) restricting access to the whole folder (with 400 permissions or similar).
In this way the risk of another user in the `fractal` group doing something on files in that folder is restricted to the time window of the workflow execution.

**Comment by tcompa** [comment]
Closing by now, in favor of #473.
